### PR TITLE
[5.1] Fix event dispatcher to allow removal of wildcard listeners

### PR DESCRIPTION
**Pull request by EliasZ**

### CODE REVIEW

**Comment by ArjanSchouten** [comment]
The same for the [```hasListeners($eventName)```](https://github.com/laravel/framework/blob/5.2/src/Illuminate/Events/Dispatcher.php#L108) method I think.

**Comment by ArjanSchouten** [comment]
I want to check the following with the ```hasListeners``` method:
```php
$d = new Dispatcher;
$this->assertFalse($d->hasListeners('foo.*'));

$d->listen('foo.*', function () { null; });
$this->assertTrue($d->hasListeners('foo.bar'));
```

That isn't working with this implementation. Think it's something to implement in a different PR @GrahamCampbell?

**Comment by taylorotwell** [comment]
@ArjanSchouten did the latest commit not address your concern?

File: tests/Events/EventsDispatcherTest.php [hunk 1]
```diff
@@ -88,6 +88,35 @@ public function testListenersCanBeRemoved()
...
+
+        $this->assertFalse(isset($_SERVER['__event.test']));
+    }
+
+    public function testListenersCanBeFound()
+    {
+        $d = new Dispatcher;
+        $this->assertFalse($d->hasListeners('foo'));
+
+        $d->listen('foo', function () { null; });
```
Comment (GrahamCampbell) [inline tests/Events/EventsDispatcherTest.php]: ?

Comment (GrahamCampbell) [inline tests/Events/EventsDispatcherTest.php]: just $d->listen('foo', function () {});

**Comment by ArjanSchouten** [comment]
In my opinion the given testcase should succeed. That's not the case with the last commit. It depends on the desired behavior and this is what I expect.

**Comment by taylorotwell** [comment]
Ah I see what you mean.

**Comment by ArjanSchouten** [comment]
The ```hasListeners``` method should try to resolve the wildcard in my opinion!

**Comment by GrahamCampbell** [comment]
Hmmm, I'd argue that's a breaking change, but I do prefer that behaviour.

**Comment by ArjanSchouten** [comment]
@GrahamCampbell Yes and no. Depends on whether you see it as a bug fix or not?

**Comment by ArjanSchouten** [comment]
Else it should go to 5.3.

**Comment by EliasZ** [comment]
For performance reasons I understand why the wildcard-events are treated as separate identifiers, but I agree with @ArjanSchouten that it would be nicer if they were treated like queries.

This way, we could also get rid of having to maintain two lists of listeners (one with wildcards and another one with regular identifiers).

[Possibly related data structure](https://github.com/laravel/framework/issues/11512#issuecomment-167218829)

**Pull request merged**